### PR TITLE
Migrate deploy + start instance modal

### LINKED ISSUE #2552: Migrate deployment+start instance modals to overlays
See [prototype](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v14/)

__What should we do?__
- Migrate the existing deployment modal to an overlay
- Explore layout design for start instance
  - Maintain 2 step approach, re-using deploy overlay
- Migrate the start instance modal to an overlay

__Why should we do it?__
It's part of our new UI architecture, related to the Visual IDE efforts.

---
Child of #2489

### PR DESCRIPTION
**Pull request by smbea**
Closes #2552 
Artifact : https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2552-migrate-deploy-modal/camunda-modeler-2552-migrate-deploy-modal-mac.dmg

- _remove fixed size from file-type icons_ -> to allow resizing to use in Platform deploy additional files.

### CODE REVIEW

**Comment by nikku** [comment]
@smbea Please consider updating this on top of https://github.com/camunda/camunda-modeler/pull/2571.

**Comment by andreasgeier** [comment]
Deployment > Add additional files

Missing container for file list with overflow:scroll. Please ensure, that the height of the file list does not exceed 3.5 lines (truncation of line 4 is intended as indicator for scrolling since the scroll bar is not shown by default).

Screen Shot 2021-11-23 at 09 44 40

Expected (see proto):

Screen Shot 2021-11-23 at 09 45 05

Note: It was not requested initially, but if the file count in the list headline (see proto) is an easy and quick addition, please decide if you can include it in this PR.

**Comment by andreasgeier** [comment]
Please ensure that long names in the file list are truncated and not expanded to new lines or break the layout. The icon size must stay stable.

Screen Shot 2021-11-23 at 09 53 05

Screen Shot 2021-11-23 at 09 55 32


Expected (see proto):

Screen Shot 2021-11-23 at 09 54 33

**Comment by andreasgeier** [comment]
Both overlays, `deploy` and `start instance with new configuration` pretend that it is possible to perform the action multiple times, which is not true.

https://user-images.githubusercontent.com/938244/143008453-a405c8bf-b725-40c8-a541-2b19eae0311f.mov

I see these options to improve:

1. The buttons stay in success state and will not be activated again as long as the overlay is opened
1. With a click on the button, the overlay closes immediately. Our existing toast message shows the success (or error) notification (as-is)
1. We believe it is useful to start multiple instances very easy by clicking multiple times on the button, so we enable this behavior for both, deploy and run.

My proposal is solution 2.

Solution 1. is a reasonable alternative.

Solution 3. would need more investigation. I can imagine that it might be a use case for run, but I don't know if it makes sense for deployment. And having two different behaviors on the overlays of these related actions should be avoided.

**Comment by smbea** [comment]
@andreasgeier The initial behaviour was to close the overlay. But since this isn't the case for any other overlay I removed this. We keep it open in the cloud overlay, for example. Would it make sense to just close these?

**Comment by pinussilvestrus** [comment]
So, solution 1 sounds not very helpful to me, because you can't actually deploy something new until you re-open. Option 2 feels more natural to me and fits what we have with the modals in the before state and also with the current execution platform selection. So I'd vote for 2) as well.

**Comment by andreasgeier** [comment]
> The initial behaviour was to close the overlay. But since this isn't the case for any other overlay I removed this.

We have different overlays in place:
- Overlays that just show information such as `What's new`. They have to be closed by clicking outside.
- Overlays that have an action button but it makes sense to keep them open for further actions. Like `Copy to Clipboard`, then open the GitHub link. So, they need to be closed by clicking outside, too.
- Overlays that have just one action like `Deploy` and `Select Platform Version` and no additional use. So we close this by clicking on the button.

Actually, the example of `Select Platform Version` is a strong argument for option 2, since it would be consistent behavior for these types of overlays.

So, from my perspective, it is okay to have different closing behavior because the use cases are different.

**Comment by nikku** [comment]
Strong preference for option :two: here :wave:

**Comment by smbea** [comment]
> * Overlays that have just one action like `Deploy` and `Select Platform Version` and no additional use. So we close this by clicking on the button.

I guess that makes sense then. Option 2 seems to make the most sense

**Comment by andreasgeier** [comment]
Agreed. Let's go with 2 and close both overlays by applying the action.

**Comment by smbea** [comment]
I updated this PR with:
* `close-on-deploy` behaviour discussed above
* changes regarding the UI of the files:

**Comment by andreasgeier** [comment]
@smbea Please provide an updated artifact.

**Comment by smbea** [comment]
@andreasgeier here you go https://camunda-modeler-on-demand.s3.eu-central-1.amazonaws.com/2552-migrate-deploy-modal/camunda-modeler-2552-migrate-deploy-modal-mac.dmg

**Comment by andreasgeier** [comment]
The problem with line break still exists if the name contains empty space:

Screen Shot 2021-11-24 at 09 23 39

I did not mentioned it explicitely, but anything that can be truncated needs a title attribute. Please add it to the list items (the file names).

**Comment by andreasgeier** [comment]
Everything else works as discussed. 

Feel free to paste a screenshot for the line break fix. Then I could give approval without a new artifact.

**Comment by smbea** [comment]
I totally missed that. Here is a new screenshot showing the line break and title tag:
Deployment

**Review by andreasgeier — APPROVED** [review]
✅ Approved. Great work, Bea!

As discussed, please let me know when this PR is merged into #2576 and I can check the styling part there.

**Comment by smbea** [comment]
**For the dev review:**
I know this is a long one (mostly because of a lot of renaming files and variables) but it can be reviewed in steps: platform then cloud, if it helps 🙂

**Comment by andreasgeier** [comment]
Ha! Sorry, right after the approval I stumbled over this bug: list item in error state is too high

Screen Shot 2021-11-24 at 10 19 19

**Comment by andreasgeier** [comment]
I added the error case to the [prototype](https://fluffy-fiesta-212e1c7c.pages.github.io/prototypes/visual-ide/v15/). The title attribute of the list item/file name is complemented by "File not found: ", meaning that the error state is not only communicated on hover of the error icon but the entire row.

Screen Shot 2021-11-24 at 11 33 40

File: client/src/plugins/camunda-plugin/start-instance-tool/StartInstanceConfigOverlay.less [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    flex-direction: column;
+  }
+
+  .intro {
+    margin: 0 0 4px 0;
+    font-size: 13px;
+    color: var(--color-grey-225-10-55);
+  }
+
+  fieldset {
```
Comment (nikku) [inline client/src/plugins/camunda-plugin/start-instance-tool/StartInstanceConfigOverlay.less:14]: Is this a generic (each form in any dialog works the same) concern? If not, why?

File: client/src/plugins/camunda-plugin/deployment-tool/DeploymentConfigOverlay.less [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    font-size: 13px;
+    color: var(--color-grey-225-10-55);
+  }
+
+  fieldset:last-child {
+    margin-bottom: 0px;
+    padding-bottom: 5px;
+  }
+
+  .overlay__footer {
```
[the remark this answers is not in the document]
Comment (nikku) [reply client/src/plugins/camunda-plugin/deployment-tool/DeploymentConfigOverlay.less:17]: Generally speaking, we shall get rid of this component entirely and rely on our newly integrated `Section` abstraction.

File: client/src/plugins/camunda-plugin/deployment-tool/DeploymentConfigOverlay.js [hunk 3]
```diff
@@ -260,21 +262,18 @@ export default class DeploymentConfigModal extends React.PureComponent {
...
                   {
                     intro && (
                       <p className="intro">
                         { intro }
                       </p>
                     )
                   }
-                  <fieldset>
+
+                  <fieldset className="fieldset">
```
Comment (nikku) [inline client/src/plugins/camunda-plugin/deployment-tool/DeploymentConfigOverlay.js:276]: Why is attaching a class / CSS override necessary here?

I'd assume we adopt a uniform way to style forms within overlays.

**Review by nikku — CHANGES_REQUESTED** [review]
Please rebase this on top of `develop` and use existing `Section` infrastructure.

Please double check the file chooser: I'd expect a scroller to show but non is present on Linux at least, cf.: 

capture GZcwIh_optimized

**Comment by smbea** [comment]
@nikku Regarding the sections: I was going to do it later on #2576, but since we are here I changed it already and got rid of the `overlay__body` and so on. To do so and still achieve the input button an a section header, I added a `relativePos` prop to be passed on to the Section, in order for me to anchor the label of the input (let me know if you see a better way to do this).

I also tried to better centralize the styles and add the scroll bar.

Since this PR is fairly big, I think we should try get this merged and further (minor of course) styling and css structure can be changed on #2576.

And regarding the errors @andreasgeier :

Pasted Graphic 3 copy

**Review by nikku — APPROVED** [review]
O.K. from my side for the moment.

Let us assess the current state next week and investigate further technical refinements once everything is in place.

**Comment by andreasgeier** [comment]
Okay for me. Let's merge and tackle styling in the other PR.

**Pull request merged**